ls1intum/Artemis
# Add token and ssh authentication possibilities to the Artemis server (GitService: clone, pull, push, ...)

### Checklist

### Motivation and Context
The Artemis server clones git repositories in many different scenarios (e.g. online code editor, creating exercises, performing empty commits, downloading repos, exporting repos, plagiarism checks). Using username and password authentication might be slower (e.g. when remote authentication services are involved) and might be less secure.

### Description
This PR adds capabilities to clone repositories using ssh (based on private-public key pairs) and private access tokens. Both should typically result in better performance and security, because the private key and token is typically longer, can be revoked more easily and the authentication check can take place directly in the version control server (Bitbucket, Gitlab) and does not need to sent to other authentication services (such as JIRA -> TUM LDAP)

**Additional changes:**
* Use unique folders for repo downloads --> should solve concurrency problems
* Improve file handling. We should not use File.separator and instead use `Paths.get(...)` or `path.resolve(subfolder)` in the future.
* Re-enable Hazelcast auto-discovery and disable TCP/IP config --> Removes errors during startup in multi-instance-setups.

### Steps for Testing
1. Log in to Artemis
2. Create a programming exercises with online code editor enabled
3. Open Edit in Editor: check that all 3 repos are cloned and displayed properly
4. Participate as student in the online code editor: check that everything is displayed properly and save and submit are working
5. Export / Download repos 
6. Execute the plagiarism check

**Additional tests:** Try out all kind of file handling activities, e.g.:
* Create drag and drop quizzes with drag item images and participate
* Upload lectures with attachments or lecture units (Attachment)
* Create file upload exercise and participate / assess / complain
* Create a course with an icon
* Upload files into markdown

### Test Coverage
no changes, the new feature cannot easily be tested automatically, existing tests have been refactored to still pass with the changes

### Screenshots
no UI changes
